Make TensorName::unifyFromRight in-place for efficiency

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29322 named tensor support for torch.equal
* **#29307 Make TensorName::unifyFromRight in-place for efficiency**

In our name inference functions we currently create an extra TensorNames
every time we unify names. This isn't completely necessary.

To do this, I made the following changes:
- TensorName now has two states, initialized and uninitialized
- Renamed unifyFromRight to unifyFromRightInplace.

Test Plan:
- `pytest test/test_namedtensor.py -v`

Differential Revision: [D18453388](https://our.internmc.facebook.com/intern/diff/D18453388)

zou3519 committed:
Make TensorName::unifyFromRight in-place for efficiency

In our name inference functions we currently create an extra TensorNames
every time we unify names. This isn't completely necessary.

To do this, I made the following changes:
- TensorName now has two states, initialized and uninitialized
- Renamed unifyFromRight to unifyFromRightInplace.

Test Plan:
- `pytest test/test_namedtensor.py -v`

[ghstack-poisoned]

nairbv commented on aten/src/ATen/TensorNames.h:
```diff
@@ -29,19 +29,25 @@ namespace at { namespace namedinference {
 // to A, `tensor` would have duplicate names [A, A]. Therefore we need to check
 // tensor.names [A, None] for the existence of A.
 struct CAFFE2_API TensorName {
+  TensorName()
```
where is this constructor needed?

zou3519 replied:
Here: https://github.com/pytorch/pytorch/pull/29307/files#diff-adc780bb266d74ff645db995562a459cR76

zou3519 replied:
`std::vector<TensorName>::resize` resizes the container. If it is resized to a larger size, then the container needs to insert new elements. The value of the new elements is either passed in via `resize`, or they are default constructed (the TensorName constructor gets called).

Now that I think about it more, there is a good way around this problem that I will go and implement to remove the `initialized_` flag. Instead of resizing the container, I can prepend to it names from the other list of names.

zou3519 replied:
I've removed the `initialized_` flag

zou3519 committed:
Update on "Make TensorName::unifyFromRight in-place for efficiency"


In our name inference functions we currently create an extra TensorNames
every time we unify names. This isn't completely necessary.

To do this, I made the following changes:
- TensorName now has two states, initialized and uninitialized
- Renamed unifyFromRight to unifyFromRightInplace.

Test Plan:
- `pytest test/test_namedtensor.py -v`

[ghstack-poisoned]

zou3519 committed:
Update on "Make TensorName::unifyFromRight in-place for efficiency"


In our name inference functions we currently create an extra TensorNames
every time we unify names. This isn't completely necessary.

To do this, I made the following changes:
- TensorName now has two states, initialized and uninitialized
- Renamed unifyFromRight to unifyFromRightInplace.

Test Plan:
- `pytest test/test_namedtensor.py -v`

[ghstack-poisoned]

nairbv commented on aten/src/ATen/TensorNames.cpp:
```diff
@@ -62,22 +62,28 @@ TensorNames::TensorNames(ArrayRef<Dimname> names, int64_t start, int64_t end) {
+  int64_t size_difference = longer_size - shorter_size;
+  bool this_is_longer = names_.size() == longer_size;
+
+  if (this_is_longer) {
```
I might prefer something like `size_diff = std::abs(names.size() - other.names.size());` and `if (names.size() > other.names.size())` if that also works, instead of tracking min/max sizes and a bool.

zou3519 replied:
Sure, that sounds reasonable

nairbv reviewed (approved):
one suggestions but this looks cleaner without the initialized tracking.

zou3519 committed:
Update on "Make TensorName::unifyFromRight in-place for efficiency"


In our name inference functions we currently create an extra TensorNames
every time we unify names. This isn't completely necessary.

To do this, I made the following changes:
- TensorName now has two states, initialized and uninitialized
- Renamed unifyFromRight to unifyFromRightInplace.

Test Plan:
- `pytest test/test_namedtensor.py -v`

[ghstack-poisoned]
